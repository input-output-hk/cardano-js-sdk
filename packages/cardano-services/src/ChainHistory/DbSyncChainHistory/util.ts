import { Cardano, PaginationArgs } from '@cardano-sdk/core';
import {
  CertificateModel,
  DelegationCertModel,
  MirCertModel,
  PoolRegisterCertModel,
  PoolRetireCertModel,
  StakeCertModel,
  WithCertType
} from './types';

export const isPoolRetireCertModel = (
  value: WithCertType<CertificateModel>
): value is WithCertType<PoolRetireCertModel> => value.type === 'retire';

export const isPoolRegisterCertModel = (
  value: WithCertType<CertificateModel>
): value is WithCertType<PoolRegisterCertModel> => value.type === 'register';

export const isMirCertModel = (value: WithCertType<CertificateModel>): value is WithCertType<MirCertModel> =>
  value.type === 'mir';

export const isStakeCertModel = (value: WithCertType<CertificateModel>): value is WithCertType<StakeCertModel> =>
  value.type === 'stake';

export const isDelegationCertModel = (
  value: WithCertType<CertificateModel>
): value is WithCertType<DelegationCertModel> => value.type === 'delegation';

export const applyPagination = (ids: Cardano.TransactionId[], { startAt, limit }: PaginationArgs) =>
  ids.slice(startAt).slice(0, limit);
