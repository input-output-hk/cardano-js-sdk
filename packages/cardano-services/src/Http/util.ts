import { ProviderError, ProviderFailure } from '@cardano-sdk/core';

export const ORIGIN = 'Origin';
export const CONTENT_TYPE = 'Content-Type';
export const APPLICATION_JSON = 'application/json';

export const getListen = (url: URL) => ({ host: url.hostname, port: Number.parseInt(url.port) });

type StaticOrigin = boolean | string | RegExp | (boolean | string | RegExp)[];

export const corsOptions = (allowedOrigins: Set<string>) => ({
  origin(requestOrigin: string | undefined, callback: (err: Error | null, options?: StaticOrigin) => void) {
    if (requestOrigin && allowedOrigins.has(requestOrigin)) {
      callback(null, requestOrigin);
    } else {
      callback(new ProviderError(ProviderFailure.Forbidden, null, `Origin ${requestOrigin} not allowed by CORS`));
    }
  }
});
