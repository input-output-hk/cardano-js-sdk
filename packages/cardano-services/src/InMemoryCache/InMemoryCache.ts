import { CACHE_TTL_DEFAULT } from './defaults';
import NodeCache from 'node-cache';

export type Key = string | number;
export type AsyncAction<T> = () => Promise<T>;

export class InMemoryCache {
  #cache: NodeCache;
  #ttlDefault: number;

  constructor(cacheTtlInMins: number = CACHE_TTL_DEFAULT, cache: NodeCache = new NodeCache()) {
    this.#ttlDefault = cacheTtlInMins * 60;
    this.#cache = cache;
  }

  /**
   * Get a cached key value, if not found execute db query and cache the result with that key
   *
   * @param key cache key
   * @param asyncAction async function to get the value
   * @param ttl cache duration in seconds
   * @returns The value stored with the key
   */
  public async get<T>(key: Key, asyncAction: AsyncAction<T>, ttl = this.#ttlDefault): Promise<T> {
    const cachedValue: T | undefined = this.#cache.get(key);
    if (cachedValue) {
      return cachedValue;
    }

    const resultPromise = asyncAction();
    this.#cache.set(
      key,
      resultPromise.catch(() => this.#cache.del(key)),
      ttl
    );
    return resultPromise;
  }

  /**
   * Get a cached key
   *
   * @param key cache key
   * @returns The value stored in the key
   */
  public getVal<T>(key: Key) {
    return this.#cache.get<T>(key);
  }

  /**
   * Set a cached key
   *
   * @param key Cache key
   * @param value A value to cache
   * @param ttl The time to live in seconds.
   */
  public set<T>(key: Key, value: T, ttl: number) {
    return this.#cache.set<T>(key, value, ttl);
  }

  /**
   * Invalidate cached values
   *
   * @param keys cache key to delete or a array of cache keys
   */
  public invalidate(keys: Key | Key[]) {
    this.#cache.del(keys);
  }

  /**
   * List all keys within this cache
   *
   * @returns An array of all keys
   */
  public keys() {
    return this.#cache.keys();
  }

  /**
   * Clear the interval timeout which is set on check period option. Default: 600
   */
  public shutdown() {
    this.#cache.close();
  }

  /**
   * Clear the whole data and reset the stats
   */
  public clear() {
    this.#cache.flushAll();
  }
}
