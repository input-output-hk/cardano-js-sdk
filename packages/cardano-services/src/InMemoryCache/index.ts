export * from './InMemoryCache';
export * from './NoCache';
export * from './defaults';
