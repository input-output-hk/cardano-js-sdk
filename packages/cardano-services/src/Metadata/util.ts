import { Cardano } from '@cardano-sdk/core';
import { TxMetadataByHashes } from './DbSyncMetadataService';
import { TxMetadataModel } from './types';
import { mapTxMetadata } from './mappers';

export const mapTxMetadataByHashes = (listOfMetadata: TxMetadataModel[]): TxMetadataByHashes => {
  const metadataMap: Map<Cardano.TransactionId, TxMetadataModel[]> = new Map();

  for (const metadata of listOfMetadata) {
    const txId = metadata.tx_id.toString('hex') as Cardano.TransactionId;
    const currentMetadata: TxMetadataModel[] = metadataMap.get(txId) ?? [];
    metadataMap.set(txId, [...currentMetadata, metadata]);
  }

  return new Map([...metadataMap].map(([id, metadata]) => [id, mapTxMetadata(metadata)]));
};
