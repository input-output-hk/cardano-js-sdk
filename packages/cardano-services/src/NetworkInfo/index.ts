export * from './DbSyncNetworkInfoProvider/index.js';
export * from './NetworkInfoHttpService.js';
