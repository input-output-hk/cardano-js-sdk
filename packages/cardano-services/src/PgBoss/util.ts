import { PgBossQueue, workerQueues } from './types';

export const isValidQueue = (queue: string): queue is PgBossQueue => workerQueues.includes(queue as PgBossQueue);

export const isErrorWithConstraint = (error: unknown): error is Error & { constraint: unknown } =>
  error instanceof Error && 'constraint' in error;
