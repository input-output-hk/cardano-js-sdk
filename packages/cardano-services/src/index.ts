export * from './Asset';
export * from './Http';
export * from './InMemoryCache';
export * from './Program';
export * from './RunnableModule';
export * from './StakePool';
export * from './TxSubmit';
export * from './Utxo';
export * from './ChainHistory';
export * from './Rewards';
export * as errors from './errors';
export * as util from './util';
