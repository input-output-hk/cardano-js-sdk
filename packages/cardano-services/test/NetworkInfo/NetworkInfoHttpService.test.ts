/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable sonarjs/no-duplicate-string */
import { Cardano, NetworkInfo, NetworkInfoProvider, testnetTimeSettings } from '@cardano-sdk/core';
import { DbSyncNetworkInfoProvider, NetworkInfoCacheKey, NetworkInfoHttpService } from '../../src/NetworkInfo';
import { HttpServer, HttpServerConfig } from '../../src';
import { InMemoryCache, UNLIMITED_CACHE_TTL } from '../../src/InMemoryCache';
import { Pool } from 'pg';
import { doServerRequest, ingestDbData, sleep, wrapWithTransaction } from '../util';
import { getPort } from 'get-port-please';
import { networkInfoHttpProvider } from '@cardano-sdk/cardano-services-client';
import axios from 'axios';

const UNSUPPORTED_MEDIA_STRING = 'Request failed with status code 415';
const APPLICATION_CBOR = 'application/cbor';
const APPLICATION_JSON = 'application/json';

describe('NetworkInfoHttpService', () => {
  let httpServer: HttpServer;
  let networkInfoProvider: DbSyncNetworkInfoProvider;
  let service: NetworkInfoHttpService;
  let port: number;
  let apiUrlBase: string;
  let config: HttpServerConfig;
  let doNetworkInfoRequest: ReturnType<typeof doServerRequest>;

  const dbPollInterval = 2 * 1000;
  const cache = new InMemoryCache(UNLIMITED_CACHE_TTL);
  const cardanoNodeConfigPath = process.env.CARDANO_NODE_CONFIG_PATH!;
  const db = new Pool({ connectionString: process.env.DB_CONNECTION_STRING, max: 1, min: 1 });

  beforeAll(async () => {
    port = await getPort();
    config = { listen: { port } };
    apiUrlBase = `http://localhost:${port}/network-info`;
    networkInfoProvider = new DbSyncNetworkInfoProvider({ cardanoNodeConfigPath, dbPollInterval }, { cache, db });
    service = await NetworkInfoHttpService.create({ networkInfoProvider });
    httpServer = new HttpServer(config, { services: [service] });
    doNetworkInfoRequest = doServerRequest(apiUrlBase);
  });

  describe('healthy state', () => {
    const dbConnectionQuerySpy = jest.spyOn(db, 'query');
    const invalidateCacheSpy = jest.spyOn(cache, 'invalidate');

    beforeEach(async () => {
      await cache.clear();
      dbConnectionQuerySpy.mockClear();
      invalidateCacheSpy.mockClear();
    });

    beforeAll(async () => {
      await httpServer.initialize();
      await httpServer.start();
    });

    afterAll(async () => {
      await db.end();
      await httpServer.shutdown();
      await cache.shutdown();
      jest.clearAllTimers();
    });

    describe('init', () => {
      it('should start epoch polling once the db provider is initialized', async () => {
        expect(cache.getVal(NetworkInfoCacheKey.CURRENT_EPOCH)).toBeUndefined();
        expect(cache.keys().length).toEqual(0);

        await sleep(dbPollInterval * 2);

        expect(cache.keys().length).toEqual(1);
        expect(cache.getVal(NetworkInfoCacheKey.CURRENT_EPOCH)).toBeDefined();
        expect(invalidateCacheSpy).not.toHaveBeenCalled();
      });
    });

    describe('/health', () => {
      it('forwards the networkInfoProvider health response', async () => {
        const res = await axios.post(`${apiUrlBase}/health`, {
          headers: { 'Content-Type': APPLICATION_JSON }
        });
        expect(res.status).toBe(200);
        expect(res.data).toEqual({ ok: true });
      });
    });

    describe('/network', () => {
      const path = '/network';
      const DB_POLL_QUERIES_COUNT = 1;
      const networkInfoTotalQueriesCount = 4;

      it('returns a 200 coded response with a well formed HTTP request', async () => {
        expect((await axios.post(`${apiUrlBase}/network`, { args: [] })).status).toEqual(200);
      });

      it('returns a 415 coded response if the wrong content type header is used', async () => {
        try {
          await axios.post(`${apiUrlBase}/network`, { args: [] }, { headers: { 'Content-Type': APPLICATION_CBOR } });
        } catch (error: any) {
          expect(error.response.status).toBe(415);
          expect(error.message).toBe(UNSUPPORTED_MEDIA_STRING);
        }
      });

      describe('cached', () => {
        it('should query the DB only once when the response is cached', async () => {
          await doNetworkInfoRequest<[], NetworkInfo>(path, []);
          await doNetworkInfoRequest<[], NetworkInfo>(path, []);

          expect(dbConnectionQuerySpy).toHaveBeenCalledTimes(networkInfoTotalQueriesCount);
          expect(cache.keys().length).toEqual(networkInfoTotalQueriesCount);
        });

        it('should call db-sync queries again once the cache is cleared', async () => {
          await doNetworkInfoRequest<[], NetworkInfo>(path, []);
          await cache.clear();
          expect(cache.keys().length).toEqual(0);

          await doNetworkInfoRequest<[], NetworkInfo>(path, []);
          expect(dbConnectionQuerySpy).toBeCalledTimes(networkInfoTotalQueriesCount * 2);
        });

        it('should not invalidate the epoch values from the cache if there is no epoch rollover', async () => {
          const currentEpochNo = 205;
          const totalQueriesCount = networkInfoTotalQueriesCount + DB_POLL_QUERIES_COUNT;

          await doNetworkInfoRequest<[], NetworkInfo>(path, []);
          expect(cache.getVal(NetworkInfoCacheKey.CURRENT_EPOCH)).toBeUndefined();
          expect(cache.keys().length).toEqual(networkInfoTotalQueriesCount);

          await sleep(dbPollInterval);

          expect(cache.getVal(NetworkInfoCacheKey.CURRENT_EPOCH)).toEqual(currentEpochNo);
          expect(cache.keys().length).toEqual(totalQueriesCount);
          expect(dbConnectionQuerySpy).toBeCalledTimes(totalQueriesCount);
          expect(invalidateCacheSpy).not.toHaveBeenCalled();
        });

        it(
          'should invalidate cached epoch values once the epoch rollover is captured by polling',
          wrapWithTransaction(async (dbConnection) => {
            const greaterEpoch = 255;

            await doNetworkInfoRequest<[], NetworkInfo>(path, []);

            await sleep(dbPollInterval);
            expect(cache.keys().length).toEqual(networkInfoTotalQueriesCount + DB_POLL_QUERIES_COUNT);

            await ingestDbData(
              dbConnection,
              'epoch',
              ['id', 'out_sum', 'fees', 'tx_count', 'blk_count', 'no', 'start_time', 'end_time'],
              [greaterEpoch, 58_389_393_484_858, 43_424_552, 55_666, 10_000, greaterEpoch, '2022-05-28', '2022-06-02']
            );

            await sleep(dbPollInterval);
            expect(invalidateCacheSpy).toHaveBeenCalledWith([
              NetworkInfoCacheKey.TOTAL_SUPPLY,
              NetworkInfoCacheKey.ACTIVE_STAKE
            ]);
            expect(cache.getVal(NetworkInfoCacheKey.CURRENT_EPOCH)).toEqual(greaterEpoch);
            expect(cache.keys().length).toEqual(3);

            await sleep(dbPollInterval);
          }, db)
        );
      });
    });

    describe('with NetworkInfoHttpProvider', () => {
      let provider: NetworkInfoProvider;

      beforeEach(async () => {
        provider = networkInfoHttpProvider(apiUrlBase);
      });

      it('time settings response is an array of network info response', async () => {
        const testnetNetworkInfo: NetworkInfo['network'] = {
          id: Cardano.NetworkId.testnet,
          magic: Cardano.CardanoNetworkMagic.Testnet,
          timeSettings: testnetTimeSettings
        };
        const response = await provider.networkInfo();
        expect(response.network).toEqual(testnetNetworkInfo);
      });

      it('response is an object of network info', async () => {
        const response = await provider.networkInfo();
        expect(response).toMatchSnapshot();
      });

      it('response is an object of ledger tip', async () => {
        const response = await provider.ledgerTip();
        expect(response).toMatchSnapshot();
      });

      it('response is an object of current wallet protocol parameters', async () => {
        const response = await provider.currentWalletProtocolParameters();
        expect(response).toMatchSnapshot();
      });

      it('response is an object of genesis parameters', async () => {
        const response = await provider.genesisParameters();
        expect(response).toMatchSnapshot();
      });
    });

    describe('/ledger-tip', () => {
      it('returns a 200 coded response with a well formed HTTP request', async () => {
        expect((await axios.post(`${apiUrlBase}/ledger-tip`, { args: [] })).status).toEqual(200);
      });

      it('returns a 415 coded response if the wrong content type header is used', async () => {
        try {
          await axios.post(`${apiUrlBase}/ledger-tip`, { args: [] }, { headers: { 'Content-Type': APPLICATION_CBOR } });
        } catch (error: any) {
          expect(error.response.status).toBe(415);
          expect(error.message).toBe(UNSUPPORTED_MEDIA_STRING);
        }
      });
    });

    describe('/current-wallet-protocol-parameters', () => {
      it('returns a 200 coded response with a well formed HTTP request', async () => {
        expect((await axios.post(`${apiUrlBase}/current-wallet-protocol-parameters`, { args: [] })).status).toEqual(
          200
        );
      });

      it('returns a 415 coded response if the wrong content type header is used', async () => {
        try {
          await axios.post(
            `${apiUrlBase}/current-wallet-protocol-parameters`,
            { args: [] },
            { headers: { 'Content-Type': APPLICATION_CBOR } }
          );
        } catch (error: any) {
          expect(error.response.status).toBe(415);
          expect(error.message).toBe(UNSUPPORTED_MEDIA_STRING);
        }
      });
    });

    describe('/genesis-parameters', () => {
      it('returns a 200 coded response with a well formed HTTP request', async () => {
        expect((await axios.post(`${apiUrlBase}/genesis-parameters`, { args: [] })).status).toEqual(200);
      });

      it('returns a 415 coded response if the wrong content type header is used', async () => {
        try {
          await axios.post(
            `${apiUrlBase}/genesis-parameters`,
            { args: [] },
            { headers: { 'Content-Type': APPLICATION_CBOR } }
          );
        } catch (error: any) {
          expect(error.response.status).toBe(415);
          expect(error.message).toBe(UNSUPPORTED_MEDIA_STRING);
        }
      });
    });
  });
});
