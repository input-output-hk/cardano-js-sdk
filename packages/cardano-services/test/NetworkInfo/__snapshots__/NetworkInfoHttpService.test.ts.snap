// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkInfoHttpService healthy state with NetworkInfoHttpProvider response is an object of current wallet protocol parameters 1`] = `
Object {
  "coinsPerUtxoWord": 34482,
  "maxCollateralInputs": 3,
  "maxTxSize": 16384,
  "maxValueSize": 5000,
  "minFeeCoefficient": 44,
  "minFeeConstant": 155381,
  "minPoolCost": 340000000,
  "poolDeposit": 500000000,
  "protocolVersion": Object {
    "major": 6,
    "minor": 0,
  },
  "stakeKeyDeposit": 2000000,
}
`;

exports[`NetworkInfoHttpService healthy state with NetworkInfoHttpProvider response is an object of genesis parameters 1`] = `
Object {
  "activeSlotsCoefficient": 0.05,
  "epochLength": 432000,
  "maxKesEvolutions": 62,
  "maxLovelaceSupply": 45000000000000000n,
  "networkId": "Testnet",
  "networkMagic": 1097911063,
  "securityParameter": 2160,
  "slotLength": 1,
  "slotsPerKesPeriod": 129600,
  "systemStart": 2019-07-24T20:20:16.000Z,
  "updateQuorum": 5,
}
`;

exports[`NetworkInfoHttpService healthy state with NetworkInfoHttpProvider response is an object of ledger tip 1`] = `
Object {
  "blockNo": 3556390,
  "hash": "332340bfcf47951b4bc8eca51c4e9190d29e2fd6dae30be231ebcdadb2d8c399",
  "slot": 58336266,
}
`;

exports[`NetworkInfoHttpService healthy state with NetworkInfoHttpProvider response is an object of network info 1`] = `
Object {
  "lovelaceSupply": Object {
    "circulating": 69165246561822582n,
    "max": 45000000000000000n,
    "total": 40408479600434778n,
  },
  "network": Object {
    "id": 0,
    "magic": 1097911063,
    "timeSettings": Array [
      Object {
        "epochLength": 21600,
        "fromSlotDate": 2019-07-24T20:20:16.000Z,
        "fromSlotNo": 0,
        "slotLength": 20000,
      },
      Object {
        "epochLength": 432000,
        "fromSlotDate": 2020-07-28T19:20:16.000Z,
        "fromSlotNo": 1598400,
        "slotLength": 1000,
      },
    ],
  },
  "stake": Object {
    "active": 15039823438862518n,
    "live": 69154436284251039n,
  },
}
`;
